agent/agent.go

```diff
@@ -3196,6 +3202,83 @@ func (a *Agent) loadProxies(conf *config.RuntimeConfig) error {
 	return persistenceErr
 }
 
+func (a *Agent) getPersistedTokens() (map[string]string, error) {
+	if !a.config.ACLEnableTokenPersistence {
+		return nil, nil
+	}
+
+	a.persistedTokensLock.RLock()
+	defer a.persistedTokensLock.RUnlock()
+
+	tokensFullPath := filepath.Join(a.config.DataDir, tokensPath)
+	tokensFile, err := os.Open(tokensFullPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("failed opening tokens file %q: %s", tokensFullPath, err)
+	}
+
+	buf, err := ioutil.ReadAll(tokensFile)
+	tokensFile.Close()
+	if err != nil {
+		return nil, fmt.Errorf("failed reading tokens file %q: %s", tokensFullPath, err)
+	}
+
+	persistedTokens := make(map[string]string)
+	if err := json.Unmarshal(buf, &persistedTokens); err != nil {
+		return nil, fmt.Errorf("failed to decode tokens file %q: %s", tokensFullPath, err)
+	}
+
+	return persistedTokens, nil
+}
+
+func (a *Agent) loadTokens(conf *config.RuntimeConfig) error {
+	persistedTokens, persistenceErr := a.getPersistedTokens()
```

Comment:
Why not hard-fail if the file is unreadable or corrupt? It seems like an unlikely event that should warrant extraordinary operational attention.

Reply:
I guess it could be a hard fail. I modeled this off of other load* functions which did similarly.

Reply:
Should a corrupt file prevent the agent from starting at all? Thats what would happen with a hard fail.

Reply:
Should corrupt persisted services prevent the agent from starting up?. I think the answer is no and we should instead just log the warning.

Reply:
If we go for worst case scenario, the agent isn't going to be terribly useful on restart if the sole store of tokens is that file, it's corrupt, and the operator is has a full belt-and-suspenders ACL setup configured.

Sure the agent will start, but it won't have any acl tokens with which to operate. If you had a fleet of hundreds of agents how can you tell the difference between an operational one and one that starts up but has no valid tokens?

But then I guess this is nearly indistinguishable from an agent having all of its acl tokens be rotated into the bit bucket already today.

Reply:
Yes, but with the agent up you can remedy the situation by PUTing new tokens to the API. Which is better than needing to, remote in, blow away/fix the corrupt file and restart (maybe still need to repost up tokens depending on the corruption).

Reply:
Well if you were being secure about it, the API would only be bound to loopback anyway, so remoting in is still required.

Reply:
I disagree that the only secure way to enable the API is bound to loopback. If using HTTPs + ACLs I would allow it on the LAN.

Additionally, even if not able to directly connect to the API I would say that doing a `consul acl set-agent-token ...` is much easier/reliable then having to hand edit files (which is still always a viable alternative).